go/tools/gopackagesdriver/flatpackage.go

```diff
@@ -0,0 +1,134 @@
+	Msg  string
+	Kind FlatPackagesErrorKind
+}
+
+type FlatPackagesErrorKind int
+
+const (
+	UnknownError FlatPackagesErrorKind = iota
+	ListError
+	ParseError
+	TypeError
+)
+
+func (err FlatPackagesError) Error() string {
+	pos := err.Pos
+	if pos == "" {
+		pos = "-" // like token.Position{}.String()
+	}
+	return pos + ": " + err.Msg
+}
+
+// FlatPackage is the JSON form of Package
+// It drops all the type and syntax fields, and transforms the Imports
+type FlatPackage struct {
+	ID              string
+	Name            string              `json:",omitempty"`
+	PkgPath         string              `json:",omitempty"`
+	Errors          []FlatPackagesError `json:",omitempty"`
+	GoFiles         []string            `json:",omitempty"`
+	CompiledGoFiles []string            `json:",omitempty"`
+	OtherFiles      []string            `json:",omitempty"`
+	ExportFile      string              `json:",omitempty"`
+	Imports         map[string]string   `json:",omitempty"`
+	Standard        bool                `json:",omitempty"`
+}
+
+type PackageFunc func(pkg *FlatPackage)
+type PathResolverFunc func(path string) string
+
+func resolvePathsInPlace(prf PathResolverFunc, paths []string) {
+	for i, path := range paths {
+		paths[i] = prf(path)
+	}
+}
+
+func WalkFlatPackagesFromJSON(jsonFile string, onPkg PackageFunc) error {
+	f, err := os.Open(jsonFile)
+	if err != nil {
+		return fmt.Errorf("unable to open package JSON file: %w", err)
+	}
+	defer f.Close()
+
+	decoder := json.NewDecoder(f)
+	for decoder.More() {
+		pkg := &FlatPackage{}
+		if err := decoder.Decode(&pkg); err != nil {
+			return fmt.Errorf("unable to decode package in %s: %w", f.Name(), err)
+		}
+		onPkg(pkg)
+	}
+	return nil
+}
+
+func (fp *FlatPackage) ResolvePaths(prf PathResolverFunc) error {
+	resolvePathsInPlace(prf, fp.CompiledGoFiles)
+	resolvePathsInPlace(prf, fp.GoFiles)
+	resolvePathsInPlace(prf, fp.OtherFiles)
+	fp.ExportFile = prf(fp.ExportFile)
+	return nil
+}
+
+func (fp *FlatPackage) IsStdlib() bool {
+	return fp.Standard
+}
+
+func (fp *FlatPackage) ResolveImports(resolve ResolvePkgFunc) {
+	// Stdlib packages are already complete import wise
+	if fp.IsStdlib() {
+		return
+	}
+
+	fset := token.NewFileSet()
+
+	for _, file := range fp.CompiledGoFiles {
+		f, err := parser.ParseFile(fset, file, nil, parser.ImportsOnly)
+		if err != nil {
+			continue
+		}
+		// If the name is not provided, fetch it from the sources
+		if fp.Name == "" {
+			fp.Name = f.Name.Name
+		}
+		for _, rawImport := range f.Imports {
+			imp, err := strconv.Unquote(rawImport.Path.Value)
+			if err != nil {
+				continue
+			}
+			// We don't handle CGo for now
+			if imp == "C" {
+				continue
+			}
+			if _, ok := fp.Imports[imp]; ok {
+				continue
+			}
+			if pkg := resolve(imp); pkg != nil {
+				if fp.Imports == nil {
+					fp.Imports = map[string]string{}
+				}
+				fp.Imports[imp] = pkg.ID
+			}
+		}
+	}
+}
+
+func (fp *FlatPackage) IsRoot() bool {
+	return strings.HasPrefix(fp.ID, "//")
```

Comment:
I don't think this is right, but I'm not sure if it matters. Maybe add a TODO? A root should be something matched directly by one of the command line arguments.

Reply:
I see. This mean I'll need to implement a matcher then.

Reply:
I have implemented a matcher, and roots should be better.